src/controllers/PartialUserFormController.php

```diff
@@ -23,76 +30,119 @@ class PartialUserFormController extends UserDefinedFormController
+    public function setData($request)
     {
         // Ensure this URL doesn't get picked up by HTTP caches
         HTTPCacheControlMiddleware::singleton()->disableCache();
 
         $key = $request->param('Key');
         $token = $request->param('Token');
 
+        /** @var PartialFormSubmission $partial */
         $partial = PartialFormSubmission::get()->find('Token', $token);
-        if (!$token || !$partial || !$partial->UserDefinedFormID) {
+        if (!$token ||
+            !$partial ||
+            !$partial->UserDefinedFormID ||
+            !hash_equals($partial->generateKey($token), $key)
+        ) {
             return $this->httpError(404);
         }
 
-        // TODO: Recognize visitor with the password
-        if ($partial->generateKey($token) === $key) {
-            // Set the session if the last session has expired
-            if (!$request->getSession()->get(PartialSubmissionController::SESSION_KEY)) {
-                $request->getSession()->set(PartialSubmissionController::SESSION_KEY, $partial->ID);
-            }
+        $session = $this->getRequest()->getSession();
+        // Set the session if the last session has expired
+        if (!$session->get(PartialSubmissionController::SESSION_KEY)) {
+            $session->set(PartialSubmissionController::SESSION_KEY, $partial->ID);
+        }
 
-            // Set data record and load the form
-            $this->dataRecord = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
-            $this->setFailover($this->dataRecord);
-
-            $form = $this->Form();
-            $fields = $partial->PartialFields()->map('Name', 'Value')->toArray();
-            $form->loadDataFrom($fields);
-
-            // Copied from {@link UserDefinedFormController}
-            if ($this->Content && $form && !$this->config()->disable_form_content_shortcode) {
-                $hasLocation = stristr($this->Content, '$UserDefinedForm');
-                if ($hasLocation) {
-                    /** @see Requirements_Backend::escapeReplacement */
-                    $formEscapedForRegex = addcslashes($form->forTemplate(), '\\$');
-                    $content = preg_replace(
-                        '/(<p[^>]*>)?\\$UserDefinedForm(<\\/p>)?/i',
-                        $formEscapedForRegex,
-                        $this->Content
-                    );
-
-                    return $this->customise([
-                        'Content' => DBField::create_field('HTMLText', $content),
-                        'Form' => '',
-                        'PartialLink' => $partial->getPartialLink()
-                    ])->renderWith([static::class, Page::class]);
-                }
-            }
+        $this->setPartialFormSubmission($partial);
+        // Set data record and load the form
+        /** @var UserDefinedForm dataRecord */
+        $this->dataRecord = DataObject::get_by_id($partial->UserDefinedFormClass, $partial->UserDefinedFormID);
```

Comment:
Any reason why the failover is not set here?

Reply:
The failover breaks for me, for some reason, and it works without...

Reply:
Without failover, the breadcrumbs are not showing

Reply:
Weird, the failover breaks everything for me

Reply:
This is what I get
![image](https://user-images.githubusercontent.com/680570/63907816-f231ab00-ca70-11e9-8807-f840988fb9de.png)